lib/Sema/CSDiag.cpp

```diff
@@ -4361,8 +4361,13 @@ bool TypeChecker::diagnoseSelfAssignment(const Expr *E) {
   if (!AE)
     return false;
 
+  auto src = AE->getSrc();
+  auto loadSrc = dyn_cast<LoadExpr>(src);
+  if (dyn_cast<ParenExpr>(loadSrc->getSubExpr()))
```

Comment:
My understanding is that the parens are an "escape hatch" - we need some way to express self-assignment in the case where it's really what you want.